src/osd/scrubber/pg_scrubber.cc

```diff
@@ -2904,43 +2904,85 @@ ostream& operator<<(ostream& out, const MapsCollectionStatus& sf)
 
-blocked_range_t::blocked_range_t(OSDService* osds,
-				 ceph::timespan waittime,
-				 ScrubMachineListener& scrubber,
-				 spg_t pg_id)
-    : m_osds{osds}
+blocked_range_t::blocked_range_t(
+    PGRef pg,
+    OSDService* osds,
+    ceph::timespan waittime,
+    ScrubMachineListener& scrubber,
+    epoch_t epoch,
+    spg_t pg_id)
+    : m_pg{pg}
+    , m_osds{osds}
     , m_scrubber{scrubber}
+    , m_epoch{epoch}
     , m_pgid{pg_id}
 {
   auto now_is = std::chrono::system_clock::now();
-  m_callbk = new LambdaContext([this, now_is]([[maybe_unused]] int r) {
-    std::time_t now_c = std::chrono::system_clock::to_time_t(now_is);
-    char buf[50];
-    strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", std::localtime(&now_c));
-    lgeneric_subdout(g_ceph_context, osd, 10)
-      << "PgScrubber: " << m_pgid
-      << " blocked on an object for too long (since " << buf << ")" << dendl;
-    m_osds->clog->warn() << "osd." << m_osds->whoami
-			 << " PgScrubber: " << m_pgid
-			 << " blocked on an object for too long (since " << buf
-			 << ")";
-
-    m_warning_issued = true;
-    m_scrubber.set_scrub_blocked(utime_t{now_c,0});
+  m_callbk = new LambdaContext([=, this]([[maybe_unused]] int r) {
+    if (m_was_deleted) {
+      return;
+    }
+    pg->lock();
+    if (!pg->pg_has_reset_since(epoch)) {
+      std::time_t now_c = std::chrono::system_clock::to_time_t(now_is);
+      char buf[50];
+      strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", std::localtime(&now_c));
+      lgeneric_subdout(g_ceph_context, osd, 10)
+	  << "PgScrubber: " << m_pgid
+	  << " blocked on an object for too long (since " << buf << ")"
+	  << dendl;
+      m_osds->clog->warn() << "osd." << m_osds->whoami
+			   << " PgScrubber: " << m_pgid
+			   << " blocked on an object for too long (since "
+			   << buf << ")";
+
+      m_warning_issued = true;
+      m_scrubber.set_scrub_blocked(utime_t{now_c, 0});
+    }
+    pg->unlock();
     return;
   });
 
   std::lock_guard l(m_osds->sleep_lock);
-  m_osds->sleep_timer.add_event_after(waittime, m_callbk);
+  m_callbk = m_osds->sleep_timer.add_event_after(waittime, m_callbk);
 }
 
+/**
+ * \attn must hold the PG lock
+ * \attn must *not* be holding the sleep_lock
+ */
+void blocked_range_t::cancel_future_alarm()
+{
+  if (!m_callbk) {
+    return;
+  }
+  if (m_warning_issued) {
+    // too late. The dtor will clear the OSD-wide flag
+    return;
+  }
+  std::lock_guard l(m_osds->sleep_lock);
+  if (m_callbk) {
+    m_osds->sleep_timer.cancel_event(m_callbk);
+    m_callbk = nullptr;
+  }
+  // we could call m_pg.reset() here (the PGRef is not needed anymore),
+  // but - as we know that the RangeBlocked state is exiting - there is
+  // no need.
+}
+
+/*
+ * we are only discarded when the RangeBlocked state is exited. Which means
+ * that we are under the PG lock.
+ */
 blocked_range_t::~blocked_range_t()
 {
+  if (!m_callbk) {
+    return;
+  }
+  m_was_deleted = true;	 // instead of accessing the sleep_timer
```

Comment:
It looks like PendingTimer owns the block_range_t instance and react(const Unblocked&)  is responsible for invoking cancel_future_alarm(), I think.

block_range_t should own invoking cancel_future_alarm.  Unblocked isn't the only way to leave RangeBlocked, and any other pathway leads to ~blocked_range_t leaking it.  I suspect the goal here is to avoid the callback invoking cancel_future_alarm().  Instead, modify the callback to clear the m_callbk member once the pg is locked and the interval check has been done and modify ~blocked_range_t to skip cancel_future_alarm() if !m_callbk.

Reply:
Not sure why this way is better. I don't consider the issue of having the callback - disabled - in the sleep queue as a major problem. It would happen only in the rare cases of blocking on an object, and discarded within a few seconds.